client/shaders/nodes_shader/opengl_fragment.glsl

```diff
@@ -153,44 +153,45 @@ float getPenumbraRadius(sampler2D shadowsampler, vec2 smTexCoord, float realDist
 	vec2 clampedpos;
 	float y, x;
 	float depth = getHardShadowDepth(shadowsampler, smTexCoord.xy, realDistance);
-	float scale_factor = 1.0;
+	float sharpness_factor = 1.0;
 	if (depth > 0.0)
-		scale_factor = clamp(depth / 1e-3, 0.0, 1.0);
-	else
-		depth = 10.0;
+		sharpness_factor = clamp(depth / 1e-3, 0.0, 1.0);
+	depth = 0.0;
 
-	if (scale_factor < 1e-2) {
+	// if we would be downscaling the filter significantly, just force a sharp shadow
+	// in practice this only happens very close to the caster and avoids peter-panning.
+	if (sharpness_factor < 1e-2) {
 		return 0.0;
 	}
 
+	float world_to_texture = xyPerspectiveBias1 / perspective_factor / perspective_factor;
+
 	if (SOFTSHADOWRADIUS <= 1.0) {
-		return scale_factor * SOFTSHADOWRADIUS;
+		return sharpness_factor * SOFTSHADOWRADIUS;
 	}
 
 
 	float pointDepth;
-	float maxRadius = SOFTSHADOWRADIUS;
+	float maxRadius = 2e-1 * SOFTSHADOWRADIUS / f_shadowfar * world_to_texture;
 	float bound = PCFBOUND;
-	float world_to_texture = xyPerspectiveBias1 / perspective_factor / perspective_factor;
-	scale_factor *= maxRadius / bound / 1024.0 * world_to_texture;
+	float scale_factor = maxRadius * sharpness_factor / bound;
 	int n = 0;
 
 	for (y = -bound; y <= bound; y += 1.0)
 	for (x = -bound; x <= bound; x += 1.0) {
 		clampedpos = vec2(x, y) * scale_factor + smTexCoord.xy;
 
 		pointDepth = getHardShadowDepth(shadowsampler, clampedpos.xy, realDistance);
-		if (pointDepth * f_shadowfar * 0.035 > length(vec2(x,y) / bound)) {
-			depth = min(depth, pointDepth);
+		if (pointDepth >= 0.0) {
+			depth += pointDepth;
+			n += 1;
 		}
 	}
 
-	if (depth == 10.0)
-		return 0.0;
+	depth *= 1e0 * world_to_texture / max(n, 1.0);
```

Comment:
ERROR[Main]: Irrlicht: 0(239) : error C1101: ambiguous overloaded function reference "max(int, float)"

Reply:
Fixed in the latest commits